### ISSUE TITLE
Regression numpy dtype incorrectly inferred when base class is unknown

### ISSUE THREAD

**Issue by NeilGirdhar** [microsoft/pyright#10185]
```python
import numpy as np
import scipy.stats as ss

# Not much I can do about this until scipy completes their distribution overhaul.
mvn: type = ss._multivariate.multivariate_normal_frozen

class A(mvn):
    pass

def f(objects: np.typing.NDArray[np.object_]):
    pass

objects = np.empty((2, 3), dtype=A)  # Has dtype float64!
# My guess is it's having trouble matching an overload.
f(objects)
```

If this is too esoteric, no problem.  If you can think of a simple workaround, pleas let me know.  Worst case, I'll switch to composition over inheritance.

I'm reporting this mainly so that you're aware of it, and in case this reveals some unintended consequence of unknown base classes.

**Comment by NeilGirdhar** [comment]
Okay, thanks for looking into it.

**Comment by jorenham** [comment]
For what it's worth; here's the `np.empty` definition:

https://github.com/numpy/numpy/blob/9389862162bbd46b5324402d6a08a27bc18ddb7d/numpy/_core/multiarray.pyi#L246-L357

**Comment by jorenham** [comment]
I can repro without scipy, and the latest stable numpy and pyright releases in a `.pyi`:

```pyi
from typing import Any
import numpy as np

obj_type: type[object]
reveal_type(np.empty(1, dtype=obj_type))        # ndarray[tuple[int], dtype[float64]]
reveal_type(np.empty((1, 2), dtype=obj_type))   # ndarray[tuple[int, int], dtype[float64]]
reveal_type(np.empty([1], dtype=obj_type))      # ndarray[tuple[int, ...], dtype[float64]]

dtype: np.dtype[np.int64]
reveal_type(np.empty(1, dtype=dtype))           # ndarray[tuple[int], dtype[float64]]
reveal_type(np.empty((1, 2), dtype=dtype))      # ndarray[tuple[int, int], dtype[float64]]
reveal_type(np.empty([1], dtype=dtype))         # ndarray[tuple[int, ...], dtype[float64]]

sctype: type[np.int64]
reveal_type(np.empty(1, dtype=dtype))           # ndarray[tuple[int], dtype[float64]]
reveal_type(np.empty((1, 2), dtype=dtype))      # ndarray[tuple[int, int], dtype[float64]]
reveal_type(np.empty([1], dtype=dtype))         # ndarray[tuple[int, ...], dtype[float64]]
```

The inferred shape-type (more of a rank-type actually) shows that the overloading seems to work correctly, or at least, for the `shape` argument.

But the inferred dtype is always `float64`, which should only happen if one the `dtype` argument is

- `None` (the default if omitted)
- `type[np.float64]`, e.g. as `dtype=np.float64`
- `np.dtype[np.float64]`, e.g. as `dtype=np.dtype(np.float64)`
- some object whose `.dtype` attribute is assignable to `np.dtype[np.float64]`

But none of these apply, and everything that is passed for `dtype` seems to be ignored.
Passing it positionally does not make a difference.

**Comment by erictraut** [comment]
My recent change exposed a different, long-standing bug that affects this code. The bug can be seen in this minimal code sample:

```python
def func(x: type):
    class A(x):
        pass

    n: None = A  # Should be an error
```

Pyright's assignability logic contained a bug that affected classes that derive from `type[Any]`. The exact metaclass of such a class cannot be determined precisely. It was not handling this case correctly. It errantly determined that the metaclass was `Any` and then concluded that an instance of this metaclass could be assigned to `None`. This caused it to pick the wrong overload for the `np.empty` call.

This will be fixed in the next release.

@Wizzerinus, this issue doesn't have any bearing on your code sample. Pyright is working as intended with your code. The issue is that `NodePath` overrides the `object.__eq__` method. When `__eq__` is overridden, it is not safe for a type checker to make assumptions about narrowing for the `in` operator, which uses the results of `__eq__` to determine containment.

**Comment by Wizzerinus** [comment]
Wow, I didn't even think of `__eq__` being the cause. Thanks for looking at it!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.399.
